src/TraceEvent/Stacks/StackSourceWriterHelper.cs

```diff
@@ -44,7 +44,11 @@ internal static class StackSourceWriterHelper
                     if (!samplesPerThread.TryGetValue(threadInfo, out var samples))
                         samplesPerThread[threadInfo] = samples = new List<Sample>();
 
-                    samples.Add(new Sample(sample.StackIndex, -1, sample.TimeRelativeMSec, sample.Metric, -1));
+                    // The single-to-double floating point error may cause unexpected overlapping of samples.
+                    // At best we can add some minimum buffer between the samples to deal with it.
```

Comment:
Overall, this change looks good.  This is the one line that gives me pause, so I'd like to just get a bit more context on it.  Is the issue that it's possible that we have a case where a sample runs up right to the next sample, (e.g. sample starts at T=0 with metric = 5, and then next sample starts at T=5), or is it something else?  I'm wondering if there should be a more robust fix where we keep track of the last sample and it's metric so that we can reliably capture these cases, and only modify the results if we have to?  This would also capture cases where there might be larger inconsistencies in the data.  What do you think?

Reply:
The real problem is with `StackSourceSample` which treats `TimeRelativeMSec` as double and `Metric` as float. The precision is already lost at

https://github.com/microsoft/perfview/blob/91a55f2e715fe1e094ad71304f413f3bf7e53ee8/src/TraceEvent/Computers/SampleProfilerThreadTimeComputer.cs#L533

when initializing the `Metric` value, and this later leads to inconsistent event sequencing as stated by https://github.com/microsoft/perfview/issues/1178. The first step to correctly manage this issue would be using a consistent data type (double) for all time values across the library. I thought it would be quite a heavy amount of work so I just made a little workaround.

If there's a better way to handle this problem or the comment can be updated with more context, I would be happy about it.

Reply:
You are correct that it would be a lot of work.  And in fact, I believe something along these lines has been attempted before and there is some subtle behavior that makes this difficult.